Add customizable log prefix and formatter to KubernetesPodOperator

HsiuChuanHsu opened the pull request:
<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->




<!-- Please keep an empty line above the dashes. -->
## Description
Addresses the issue where KubernetesPodOperator (KPO) logs always include a container name prefix, which can consume unnecessary horizontal space and feel redundant. This PR introduces two new optional parameters to `KubernetesPodOperator`:
- `log_prefix`: A boolean to enable/disable the container name prefix in logs (default: True).
- `log_formatter`: A callable to fully customize log output format, overriding log_prefix if provided.

This allows users to either disable the prefix or define a custom log format, improving log readability and flexibility.

## Changes
- Added `log_prefix: bool = True` and `log_formatter: Callable[[str, str], str] | None = None` to `KubernetesPodOperator` constructor.
- Updated `PodManager` methods (`fetch_container_logs`, `fetch_requested_init_container_logs`, `fetch_requested_container_logs`) to accept and use these new parameters.
- Modified log handling in `PodManager.fetch_container_logs` to:
  - Apply `log_formatter` if provided.
  - Use `log_prefix` to include/exclude container name prefix when `log_formatter` is not set.
- Added pytest unit tests to verify:
  - Default behavior with `log_prefix=True`.
  - Behavior with `log_prefix=False` (no container name prefix).
  - Custom log formatting with `log_formatter`.

## Related Issue
#53054 

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [airflow-core/newsfragments](https://github.com/apache/airflow/tree/main/airflow-core/newsfragments).

HsiuChuanHsu committed:
Add custom log formatting and prefix control to KubernetesPodOperator

- Introduce  parameter to toggle container name prefix in logs
- Add  parameter for custom log message formatting
- Update PodManager to support log_prefix and log_formatter in log fetching

HsiuChuanHsu committed:
Add tests for log prefix and custom formatter in KubernetesPodOperator

- Add test_log_prefix_enabled to verify default log_prefix=True behavior
- Add test_log_prefix_disabled to confirm log_prefix=False removes container prefix
- Add test_custom_log_formatter to validate custom log_formatter functionality

jason810496 commented on kubernetes-tests/tests/kubernetes_tests/test_kubernetes_pod_operator.py:
```diff
@@ -1466,6 +1466,74 @@ def test_init_container_logs_filtered(self, mock_get_connection):
             < calls_args.find(marker_from_main_container)
         )
 
+    @pytest.mark.asyncio
```
How about combine these 3 test scenarios with 1 test method with `pytest.mark.parameterize` ?
Since the only difference among the tests are the `log_prefix`, `log_formatter` setup and expected value in assertion.
Just personal preference, no strong opinion for this.

HsiuChuanHsu replied:
Sure, I'll combine the three test scenarios into one test method using `pytest.mark.parametrize`, as they only differ in `log_prefix`, `log_formatter setup`, and the expected assertion value. 
Thanks for the suggestion. I'll get to work on it!

jason810496 reviewed (commented):
Thanks for the PR! LGTM overall but the unit still needs fix.

HsiuChuanHsu committed:
Enhance KubernetesPodOperator logging with flexible prefix and formatter options

Consolidated log prefix and formatter tests into a single parameterized test_log_output_configurations, covering:
log_prefix=True: Includes container name prefix ([base]).
log_prefix=False: Excludes container name prefix.
Custom log formatter: Applies user-defined formatting (e.g., CUSTOM[base]: message)

jason810496 commented on kubernetes-tests/tests/kubernetes_tests/test_kubernetes_pod_operator.py:
```diff
@@ -1466,6 +1467,56 @@ def test_init_container_logs_filtered(self, mock_get_connection):
             < calls_args.find(marker_from_main_container)
         )
 
+    @pytest.mark.asyncio
```
```suggestion
```

HsiuChuanHsu replied:
Thanks for the review! The @pytest.mark.asyncio is removed.

jason810496 reviewed (approved):
Thanks for the update, only a small nit and this PR is good to be merged.

After #53475 has merged, we don't need `pytest.mark.asyncio` any more.
It would be nice to remove the mark and rebase again, thanks!

HsiuChuanHsu committed:
Remove @pytest.mark.asyncio

Co-authored-by: LIU ZHE YOU <68415893+jason810496@users.noreply.github.com>

jason810496 commented on kubernetes-tests/tests/kubernetes_tests/test_kubernetes_pod_operator.py:
```diff
@@ -1428,6 +1429,55 @@ def test_init_container_logs_filtered(self, mock_get_connection):
+        ],
+    )
+    def test_log_output_configurations(
+        self, mock_get_connection, caplog, log_prefix_enabled, log_formatter, expected_log_message_check
```
Sorry, I miss the `caplog` usage in last review.

Since `caplog` introduce some flakiness in the tests, so we need to replace `caplog` with mock.

Or the static check in CI will fail:

https://github.com/apache/airflow/actions/runs/16552432381/job/46809215628

HsiuChuanHsu replied:
I have noticed the failed CI, but i have no idea how to fix the error in the first place.😭 I think i will figure this out!

HsiuChuanHsu committed:
Modify test_log_output_configurations to improve log message capturing:
    - Replace caplog with targeted mocking of PodManager logger

HsiuChuanHsu committed:
Add custom log formatting support to KubernetesPodOperator tests

jason810496 commented on providers/cncf/kubernetes/src/airflow/providers/cncf/kubernetes/operators/pod.py:
```diff
@@ -343,6 +343,8 @@ def __init__(
         progress_callback: Callable[[str], None] | None = None,
         logging_interval: int | None = None,
         trigger_kwargs: dict | None = None,
+        log_prefix: bool = True,
```
Would it be better to rename `log_prefix` to `container_name_log_prefix_enabled`?
Since the purpose of the `log_prefix` flag is to control whether the `[container_name]` prefix appears in the logs or not.

HsiuChuanHsu replied:
True! container_name_log_prefix_enabled would be more understandable.
Fixed!

jason810496 commented on providers/cncf/kubernetes/src/airflow/providers/cncf/kubernetes/operators/pod.py:
```diff
@@ -343,6 +343,8 @@ def __init__(
         logging_interval: int | None = None,
         trigger_kwargs: dict | None = None,
+        log_prefix: bool = True,
+        log_formatter: Callable[[str, str], str] | None = None,
```
The docstring of `container_name_log_prefix_enabled` and `log_formatter` is missing.

https://github.com/apache/airflow/blob/16ee837a91cf9ee48c1596197b4df41bcb46ee29/providers/cncf/kubernetes/src/airflow/providers/cncf/kubernetes/operators/pod.py#L237-L238

Additionally, it would be nice to describe the type of `log_formatter` in the docstring, as first str is `container_name` and second str is `message_to_log`.

HsiuChuanHsu replied:
Oh.. I missed this part. Thanks for the reminder!
Fixed!

jason810496 commented on providers/cncf/kubernetes/src/airflow/providers/cncf/kubernetes/utils/pod_manager.py:
```diff
@@ -532,7 +534,17 @@ def consume_logs(*, since_time: DateTime | None = None) -> tuple[DateTime | None
+                                                if log_prefix
+                                                else message_to_log
+                                            )
+                                            self.log.info("%s", log_message)
```
It seems we can add a common function to avoid duplicate logic.

HsiuChuanHsu replied:
Agree! Make it into function `_log_message`.

jason810496 commented on providers/cncf/kubernetes/src/airflow/providers/cncf/kubernetes/utils/pod_manager.py:
```diff
@@ -551,7 +563,14 @@ def consume_logs(*, since_time: DateTime | None = None) -> tuple[DateTime | None
+                                log_message = (
+                                    f"[{container_name}] {message_to_log}" if log_prefix else message_to_log
+                                )
+                                self.log.info("%s", log_message)
```
Since the above logic is exactly same here.

HsiuChuanHsu replied:
Fixed!

jason810496 reviewed (commented):
Nice! Thanks for the fix!

HsiuChuanHsu committed:
feat(kubernetes): Add log formatting options and refactor log handling in KubernetesPodOperator

- Rename log_prefix to  container_name_log_prefix_enabled parameter to control container name prefix in logs

- Extract duplicate log formatting logic into _log_message private method in PodManager

- Update docstrings with detailed parameter descriptions
